### PR TITLE
Add instance_id for gce imported hosts

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
instance_id is a feature with inventory imports to track the identity of hosts.

This adds it for gce, it's already a thing for all other major types

##### ISSUE TYPE
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
4.0.0
```

##### ADDITIONAL INFORMATION
Revives https://github.com/ansible/awx/pull/3362

Connect #3141

### CODE REVIEW

File: awx/main/migrations/0069_v350_gce_instance_id.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+# new value introduced in awx/settings/defaults.py, frozen in time here
+GCE_INSTANCE_ID_VAR = 'gce_id'
+
+
+def gce_id_forward(apps, schema_editor):
+    set_new_instance_id(apps, 'gce', GCE_INSTANCE_ID_VAR)
+
+
+def gce_id_backward(apps, schema_editor):
```
Comment (ryanpetrello) [inline awx/main/migrations/0069_v350_gce_instance_id.py]: Migration rollbacks are a lie :)

File: awx/main/migrations/_inventory_source.py [hunk 2]
```diff
@@ -62,3 +65,84 @@ def remove_azure_inventory_sources(apps, schema_editor):
...
+    return new_id_value
+
+
+def set_new_instance_id(apps, source, new_id):
+    '''This methods adds an instance_id in cases where there was not one before
+    '''
+    from django.conf import settings
+    id_from_settings = getattr(settings, '{}_INSTANCE_ID_VAR'.format(source.upper()))
+    if id_from_settings != new_id:
+        # User applied an instance ID themselves, so nope on out of there
```
Comment (ryanpetrello) [inline awx/main/migrations/_inventory_source.py:111]: Yea, I'm fine with this 👍

File: awx/main/migrations/_inventory_source.py [hunk 3]
```diff
@@ -62,3 +65,84 @@ def remove_azure_inventory_sources(apps, schema_editor):
     logger.debug("Removing all Azure InventorySource from database.")
     InventorySource.objects.filter(source='azure').delete()
 
+
+def _get_instance_id(from_dict, new_id, default=''):
+    '''logic mostly duplicated with inventory_import command Command._get_instance_id
+    frozen in time here, for purposes of migrations
+    '''
+    instance_id = default
+    for key in new_id.split('.'):
```
Comment (ryanpetrello) [inline awx/main/migrations/_inventory_source.py:74]: (for reference for other PR readers)

https://github.com/ansible/awx/blob/devel/awx/main/management/commands/inventory_import.py#L273

**Review by ryanpetrello — APPROVED** [review]

File: awx/main/migrations/_inventory_source.py [hunk 4]
```diff
@@ -62,3 +65,84 @@ def remove_azure_inventory_sources(apps, schema_editor):
...
+        from_dict = instance_id
+    return smart_text(instance_id)
+
+
+def _get_instance_id_for_upgrade(host, new_id):
+    if host.instance_id:
+        # this should not have happened, but nothing to really do about it...
+        logger.debug('Unexpectedly, host {}-{} has instance_id set'.format(host.name, host.pk))
+        return None
+    host_vars = parse_yaml_or_json(host.variables)
```
Comment (ryanpetrello) [inline awx/main/migrations/_inventory_source.py:88]: For large inventories, I _could_ see this being a potentially time consuming thing if you've got a lost of hostvars.  This is potentially tens of thousands of "parse a large-ish JSON document, compare things, save a new value to the database."

Have you done any timing of this with large, fake datasets (i.e., a few thousand hosts w/ really large JSON blobs)?

**Comment by AlanCoding** [comment]
Rebased branch so that migration would be bumped.

File: awx/main/migrations/_inventory_source.py (hunk 4 above)
Comment (AlanCoding) [reply awx/main/migrations/_inventory_source.py:88]: This is the reason I like the reverse migration.

https://gist.github.com/AlanCoding/40a9209dac5cc065c65faa673229127b

I can dive into speculation with a few specifics from this. I have 166 inventories with the same gce source imported redundantly. If you do the division, that's a dozen hosts that are imported from this import.

Either we are worried about large single gce sources, or we are worried about a large number of gce sources. In the case of the former, we already do parsing of the JSON for all hosts in the inventory import, so as long as we're not _worse_ than that, doesn't really matter. The former is more murky. Sure 166x12 isn't huge, we might be more worried about 30x5,000. If the concern is migration time, the current data seems relatively reassuring. Memory could be a genuine problem if some things do not get garbage collected.

For the size of the variables... these tend to be 1,000 characters (as the JSON formatted string) before the inventory plugin, and 4,000 characters after. We will only be migrating the pre-plugin versions of these hosts (except for a very small population of AWX users).

**Pull request merged**